WalletWasabi/Blockchain/TransactionOutputs/CoinsRegistry.cs

```diff
@@ -118,16 +118,17 @@ private bool TryAddNoLock(SmartCoin coin)
 			{
 				hashSet = new();
 				CoinsByTransactionId.Add(coin.TransactionId, hashSet);
+
+				// Each prevOut of the transaction contributes to the existence of coins.
+				// This only has to be added once per transaction.
```

Comment:
Here the reason is that we will call this loop once per output, but each output references the same object `SmartTransaction`, so its inputs will be the same at each iteration.

I agree that it would be better to throw here. The reason I'm not doing that is to be safe and to follow the legacy behaviour of this cache, as it was already working like that prior to our recent changes to the `CoinsRegistry`. I have to think a bit more about the implications of throwing here. Also, why `throw` and don't call `Undo`?

Reply:
I added that and it makes the tests to fail. Do you mind if we do this in another PR?